app/components/Approvals/Snaps/SnapDialogApproval/SnapsDialogApproval.tsx

```diff
@@ -0,0 +1,152 @@
+///: BEGIN:ONLY_INCLUDE_IF(preinstalled-snaps)
+import React, { useState } from 'react';
+import { View } from 'react-native';
+import { useStyles } from '../../../hooks/useStyles';
+import { strings } from '../../../../../locales/i18n';
+import stylesheet from './SnapDialogApproval.styles';
+import useApprovalRequest from '../../../Views/confirmations/hooks/useApprovalRequest';
+import ApprovalModal from '../../ApprovalModal';
+import BottomSheetFooter, {
+  ButtonsAlignment,
+} from '../../../../component-library/components/BottomSheets/BottomSheetFooter';
+import {
+  ButtonVariants,
+  ButtonSize,
+} from '../../../../component-library/components/Buttons/Button';
+import Engine from '../../../../core/Engine';
+import { SnapUIRenderer } from '../SnapUIRenderer/SnapUIRenderer';
+import { Json, SnapId } from '@metamask/snaps-sdk';
+import { DIALOG_APPROVAL_TYPES } from '@metamask/snaps-rpc-methods';
+
+export enum TemplateConfirmation {
+  Ok = 'template_confirmation.ok',
+  CANCEL = 'template_confirmation.cancel',
+}
+
+const SnapDialogApproval = () => {
+  const [isLoading, setIsLoading] = useState(false);
+  const { approvalRequest } = useApprovalRequest();
+  const { styles } = useStyles(stylesheet, {});
+
+  const onCancel = async () => {
+    if (!approvalRequest) return;
+    await Engine.acceptPendingApproval(
+      approvalRequest.id,
+      null as unknown as Record<string, Json>,
+    );
+  };
+
+  const onConfirmInput = async () => {
+    setIsLoading(true);
+    if (!approvalRequest) return;
+
+    const inputState =
+      await Engine.context.SnapInterfaceController.getInterface(
+        approvalRequest?.origin as SnapId,
+        approvalRequest.requestData.id,
+      );
+    await Engine.acceptPendingApproval(
+      approvalRequest.id,
+      inputState.state['custom-input'] as unknown as Record<string, Json>,
+    );
+    setIsLoading(false);
+  };
+
+  const onConfirm = async () => {
+    setIsLoading(true);
+    if (!approvalRequest) return;
+    await Engine.acceptPendingApproval(
+      approvalRequest.id,
+      true as unknown as Record<string, Json>,
+    );
+
```

Comment:
We should be calling `SnapInterfaceController.deleteInterface(approvalRequest.id)` when confirming and cancelling an approval

Reply:
How does that work in this case when we're returning the `inputState.state['custom-input'] as unknown as Record<string, Json>` back to the user? Seeing as deleteInterface only seems to accept the request id and not any other parameters.

Reply:
```
  const inputState =
      await Engine.context.SnapInterfaceController.getInterface(
        approvalRequest?.origin as SnapId,
        approvalRequest.requestData.id,
      );
    await Engine.acceptPendingApproval(
      approvalRequest.id,
      inputState.state['custom-input'] as unknown as Record<string, Json>,
    );
```
Here we get the current input state and return back to the user `onConfirm`. The input state is then displayed in the grey field on the Snaps dialog screen:

<img width="363" alt="Screenshot 2025-02-06 at 15 36 12" src="https://github.com/user-attachments/assets/7b532bae-2c72-4e40-a84e-1c2d1b2a1361" />

Reply:
We should continue to resolve the pending approval, we just need to clean up and delete the interface afterwards!

Reply:
I suspect this hasn't been configured correctly in the `Engine.ts` file so I'm going to dig into the implementation on extension first.

Reply:
We do the same on extension, see here for example: https://github.com/MetaMask/metamask-extension/blob/fb/snap-ui-renderer-tests/ui/pages/confirmations/confirmation/templates/snaps/snap-confirmation/snap-confirmation.js#L10

Reply:
So the extension implementation is different of course, but does this look correct to you? This is the `acceptApproval` function:

```
async acceptPendingApproval(
    id: string,
    requestData?: Record<string, Json>,
    opts: AcceptOptions & { handleErrors?: boolean } = {
      waitForResult: false,
      deleteAfterResult: false,
      handleErrors: true,
    },
  ) {
    const { ApprovalController } = this.context;

    try {
      return await ApprovalController.accept(id, requestData, {
        waitForResult: opts.waitForResult,
        deleteAfterResult: opts.deleteAfterResult,
      });
    } catch (err) {
      if (opts.handleErrors === false) {
        throw err;
      }
    }
  }
```

Hard to tell, but it feels kind of close to the `deleteInterface` function.

Reply:
That looks fine, my only point is that we should also do something like this afterwards:
```suggestion
Engine.context.SnapInterfaceController.deleteInterface(
  approvalRequest.requestData.id,
)
```

Reply:
Had a look at the `ApprovalController` types and it seems `accept` deletes the approval, which I'm assuming `deleteInterface` also does? Also pasting `reject` here so you can see the difference between them.

```
accept(id: string, value?: unknown, options?: AcceptOptions): Promise<AcceptResult>;
    /**
     * Rejects the promise of the approval with the given id, and deletes the
     * approval. Throws an error if no such approval exists.
     *
     * @param id - The id of the approval request.
     * @param error - The error to reject the approval promise with.
     */
    reject(id: string, error: unknown): void;
    /**
     * Rejects and deletes all approval requests.
     *
     * @param rejectionError - The JsonRpcError to reject the approval
     * requests with.
     */
```

Reply:
Yep! The Snap interface is stored separately from the approval, so we need to clean up both.

E.g. the approval simply stores an identifier for which interface to show.